### PR TITLE
Reimplementation

### PR DESCRIPTION
**Pull request by buchi**
- Not compatible with ftw.solr 1.x, currently there's no upgrade path
- No longer depends on collective.solr
- Provides it's own Solr integration using Solr's REST API
- Does not try to replace the portal_catalog and doesn't patch the catalogs search method
- Automatically indexes all Plone objects in Solr
- Does not provide a search integration into Plone's default search form
- Has to be used explicitly for searching in custom code
- Works with latest Apache Solr 7.2.1

### CODE REVIEW

File: ftw/solr/profiles/default/componentregistry.xml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0"?>
+<componentregistry>
+ <adapters/>
+ <utilities>
+  <utility
```
Comment (deiferni) [inline ftw/solr/profiles/default/componentregistry.xml]: this will be a persistent utility? why is this necessary? couldn't you just register this thing with `zcml`?

File: ftw/solr/__init__.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+# -*- coding: utf-8 -*-
 from zope.i18nmessageid import MessageFactory
 
+
 _ = MessageFactory('ftw.solr')
```
Comment (deiferni) [inline ftw/solr/__init__.py]: i could not see any translations, drop the messagefactory?

File: ftw/solr/locales/update.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (deiferni) [inline ftw/solr/locales/update.sh]: there is already an `bin/i18n-build` provided by development buildout. i think this script is not necessary.
also, there are longer any translations, so the locales directory could be removed?

File: ftw/solr/handlers.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+from plone.indexer.interfaces import IIndexableObject
+from plone.namedfile.interfaces import INamedBlobFile
+from plone.rfc822.interfaces import IPrimaryFieldInfo
+from zope.component import queryMultiAdapter
+from zope.interface import implementer
+import logging
+
+
+logger = getLogger('ftw.solr.handlers')
+logger.setLevel(logging.DEBUG)
```
Comment (deiferni) [inline ftw/solr/handlers.py]: so this will log in debug level by default? is this necessary?

File: ftw/solr/setuphandlers.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+class HiddenProfiles(object):
+
+    def getNonInstallableProfiles(self):
+        """Hide uninstall profile from site-creation and quickinstaller."""
+        return [
+            'ftw.solr:uninstall',
+        ]
+
+
+def post_install(context):
```
Comment (deiferni) [inline ftw/solr/setuphandlers.py]: this seems unused. remove?

File: ftw/solr/setuphandlers.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    """Post install script"""
+    # Do something at the end of the installation of this package.
+
+
+def uninstall(context):
```
Comment (deiferni) [inline ftw/solr/setuphandlers.py]: dito.

File: ftw/solr/__init__.py (hunk 2 above)
Comment (buchi) [reply ftw/solr/__init__.py]: I've removed all translations related stuff as there are indeed no translations.

File: ftw/solr/connection.py [hunk 7]
```diff
@@ -157,20 +160,57 @@ def schema(self):
...
+        self.start = 0
+        self.body = {}
+        if self.http_status:
+            self.parse(body)
+
+    def parse(self, data):
+        try:
+            data = json.loads(data)
+        except ValueError:
+            data = None
```
[the remark this answers is not in the document]
Comment (buchi) [reply ftw/solr/connection.py]: Invalid JSON response, e.g. when calling a wrong url and the server returns HTML.
An error is already logged, see https://github.com/4teamwork/ftw.solr/blob/tb-reimplementation/ftw/solr/connection.py#L197

File: ftw/solr/connection.py [hunk 8]
```diff
@@ -157,20 +160,57 @@ def schema(self):
...
+            self.docs = data[u'response'].get(u'docs', [])
+            self.num_found = data[u'response'].get(u'numFound', 0)
+            self.start = data[u'response'].get(u'start', 0)
+
+        self.body = data
 
     def is_ok(self):
-        if self.status == 200 and self.get(u'responseHeader', {}).get(
-                u'status') == 0:
+        if self.http_status == 200 and self.status == 0:
```
[the remark this answers is not in the document]
Comment (buchi) [reply ftw/solr/connection.py]: IMHO the current implementation is more readable

File: ftw/solr/handlers.py (hunk 4 above)
Comment (buchi) [reply ftw/solr/handlers.py]: That was just for debugging purpose. Removed.

File: ftw/solr/locales/update.sh (hunk 3 above)
Comment (buchi) [reply ftw/solr/locales/update.sh]: removed

File: ftw/solr/connection.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (deiferni) [reply ftw/solr/connection.py]: 👍

File: ftw/solr/connection.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (deiferni) [reply ftw/solr/connection.py]: fair enouh 😉

File: ftw/solr/setuphandlers.py (hunk 5 above)
Comment (buchi) [reply ftw/solr/setuphandlers.py]: removed

File: ftw/solr/setuphandlers.py (hunk 6 above)
Comment (buchi) [reply ftw/solr/setuphandlers.py]: removed

File: ftw/solr/document.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        path = self.getPath()
+        if not path:
+            return None
+        path = path.split('/')
+        if restricted:
+            parent = site.unrestrictedTraverse(path[:-1], None)
+            if parent is None:
+                return None
+            return parent.restrictedTraverse(path[-1], None)
+        return site.unrestrictedTraverse(path, None)
```
Comment (jone) [inline ftw/solr/document.py]: When something is outdated it could happen that we get an object here, but it's path is not `path`.
This can happen when there is no object at `path`, but there is an uncle with the same id - acquisition will return another object then.
Should we handle this case by checking that the traversed object actually has the path we expect it to have?

I ran into this problem with the catalog now and then..

Comment (buchi) [reply ftw/solr/document.py]: Good point. I've added a check to make sure we only return an object if it's path matches.
